sklearn/linear_model/ridge.py

```diff
@@ -545,6 +545,13 @@ def fit(self, X, y, sample_weight=None):
                          accept_sparse=_accept_sparse,
                          dtype=_dtype,
                          multi_output=True, y_numeric=True)
+        if sparse.issparse(X) and self.fit_intercept:
+            solver = 'sparse_cg'
+            if self.solver not in ['auto', 'sparse_cg']:
```

Comment:
The solver resolution (e.g. for `auto` is normally done in `_ridge_regression`) maybe better to move this there as well.

Reply:
The difficulty is that depending on `fit_intercept` and whether `X` is dense we
need to provide `_ridge_regression` with `X_offset` and `X_scale` (computed in
preprocessing) or not:

https://github.com/scikit-learn/scikit-learn/blob/e871a56d445183503e64e7b23bff43c4acd30693/sklearn/linear_model/ridge.py#L571

Reply:
I would not have prevented someone to use 'sag' with fit_intercept=True. It's not broken per se it is just that it needs a lore more iterations than the default value.

Reply:
In this case should there be a warning? Users may be surprised by the number of
iterations they need to set:

```python
>>> x, y = _make_sparse_offset_regression(n_samples=20, n_features=5, random_state=0)                                                             
>>> sp_ridge = Ridge(solver='sag', max_iter=10000000, tol=1e-8).fit(sparse.csr_matrix(x), y)                                                      
>>> ridge = Ridge(solver='sag', max_iter=10000000, tol=1e-8).fit(x, y)                                                                            
>>> sp_ridge.n_iter_[0]                                                                                                                           
566250
>>> ridge.n_iter_[0]                                                                                                                              
100
>>> np.allclose(sp_ridge.intercept_, ridge.intercept_, rtol=1e-3)                                                                                 
False
```

Reply:
@agramfort I restored the possibility of using 'sag' and added a warning, let me know if I should remove the warning